gateway/handler.go

```diff
@@ -728,6 +734,50 @@ func handleServiceWorkerRegistration(r *http.Request) (err *ErrorResponse) {
 	return nil
 }
 
+// handleIpnsB58mhToCidRedirection redirects from /ipns/b58mh to /ipns/cid in
+// the most cost-effective way.
+func handleIpnsB58mhToCidRedirection(w http.ResponseWriter, r *http.Request) bool {
+	if w.Header().Get("Location") != "" {
+		// Ignore this if there is already a redirection in place. This happens
+		// if there is a subdomain redirection. In that case, the path is already
+		// converted to CIDv1.
+		return false
+	}
+
+	pathParts := strings.Split(r.URL.Path, "/")
+	if len(pathParts) < 3 {
+		return false
+	}
+
+	if pathParts[1] != "ipns" {
+		return false
+	}
+
+	id, err := peer.Decode(pathParts[2])
+	if err != nil {
+		return false
+	}
+
+	// Convert the peer ID to a CIDv1.
+	cid := peer.ToCid(id)
+
+	// Encode CID in base36 to match the subdomain URLs.
+	encodedCID, err := cid.StringOfBase(multibase.Base36)
+	if err != nil {
+		return false
+	}
+
+	// If the CID was already encoded, do not redirect.
+	if encodedCID == pathParts[2] {
+		return false
+	}
+
+	pathParts[2] = encodedCID
+	r.URL.Path = strings.Join(pathParts, "/")
+	http.Redirect(w, r, r.URL.String(), http.StatusFound)
```

Comment:
@hacdias  I slightly worry this will break  DNSLink websites that have DNS TXT record with value `dnslink=/ipns/Qm..`   or `dnslink=/ipns/1..`

In such situations, `example.net` being backed by `/ipns/id` would redirect to `example.net/ipns/base36-id` which will return 404.

Mind adding a check that redirects ONLY when  `DNSLinkHostnameKey` is missing?

Reply:
Done!